data/src/main/java/org/apache/iceberg/data/GenericReader.java

```diff
@@ -45,14 +45,14 @@
 import org.apache.iceberg.util.PartitionUtil;
 
 class GenericReader implements Serializable {
-  private final FileIO io;
+  private final EncryptingFileIO io;
   private final Schema tableSchema;
   private final Schema projection;
   private final boolean caseSensitive;
   private final boolean reuseContainers;
 
   GenericReader(TableScan scan, boolean reuseContainers) {
-    this.io = scan.table().io();
+    this.io = EncryptingFileIO.combine(scan.table().io(), scan.table().encryption());
```

Comment:
the table.io() comes from the catalog's table operations, and will already be an EncryptingFileIO for encrypted tables.